### PR TITLE
CRI-O first integration

### PR DESCRIPTION
**Pull request by giuseppe**
based on https://github.com/openshift/openshift-ansible/pull/4763

### CODE REVIEW

**Comment by ashcrow** [comment]
```
./roles/openshift_cli/library/openshift_container_binary_sync.py:45:5: E303 too many blank lines (2)
./roles/openshift_cli/library/openshift_container_binary_sync.py:52:5: E303 too many blank lines (2)
./roles/openshift_cli/library/openshift_container_binary_sync.py:72:5: E303 too many blank lines (2)
./roles/openshift_cli/library/openshift_container_binary_sync.py:102:5: E303 too many blank lines (2)
./roles/openshift_cli/library/openshift_container_binary_sync.py:116:5: E303 too many blank lines (2)
```

```
roles/openshift_cli/library/openshift_container_binary_sync.py:57: [C0301(line-too-long), ] Line too long (128/120)
roles/openshift_cli/library/openshift_container_binary_sync.py:28: [R0902(too-many-instance-attributes), BinarySyncer] Too many instance attributes (8/7)
```

**Comment by giuseppe** [comment]
@ashcrow thanks for the quick review, I've pushed a new version that solves these issues

**Comment by ashcrow** [comment]
```
roles/openshift_cli/library/openshift_container_binary_sync.py:55: [C0301(line-too-long), ] Line too long (128/120)
roles/openshift_cli/library/openshift_container_binary_sync.py:28: [R0902(too-many-instance-attributes), BinarySyncer] Too many instance attributes (8/7)
```

**Comment by ashcrow** [comment]
@giuseppe it looks like ``use_crio`` isn't set yet when installing. The following should get around that:

```diff
diff --git a/roles/openshift_facts/tasks/main.yml b/roles/openshift_facts/tasks/main.yml
index c0d3e78..a720184 100644
--- a/roles/openshift_facts/tasks/main.yml
+++ b/roles/openshift_facts/tasks/main.yml
@@ -7,6 +7,7 @@
 # Locally setup containerized facts for now
 - set_fact:
     l_is_atomic: "{{ ostree_booted.stat.exists }}"
+    l_use_crio: "{{ openshift_docker_use_crio }}"
 - set_fact:
     l_is_containerized: "{{ (l_is_atomic | bool) or (containerized | default(false) | bool) }}"
     l_is_openvswitch_system_container: "{{ (openshift_use_openvswitch_system_container | default(openshift_use_system_containers) | bool) }}"
@@ -55,7 +56,7 @@
       - l_atomic_docker_version.stdout | replace('"', '') | version_compare('1.12','>=')
 
   when:
-  - not use_crio
+  - not l_use_crio
   - l_is_atomic | bool
   - r_openshift_facts_ran is not defined
```

**Comment by giuseppe** [comment]
thanks!  Amended and pushed a new version

**Comment by ashcrow** [comment]
@giuseppe I think it needs a default value of ``False`` as well.

**Comment by ashcrow** [comment]
Similar issue but in ```openshift_cli```

```diff
diff --git a/roles/openshift_cli/tasks/main.yml b/roles/openshift_cli/tasks/main.yml
index 1a211ef..45b8983 100644
--- a/roles/openshift_cli/tasks/main.yml
+++ b/roles/openshift_cli/tasks/main.yml
@@ -1,4 +1,7 @@
 ---
+- set_fact:
+    l_use_crio: "{{ openshift_docker_use_crio | default(false) }}"
+
 - name: Install clients
   package: name={{ openshift.common.service_type }}-clients state=present
   when: not openshift.common.is_containerized | bool
@@ -17,7 +20,7 @@
       backend: "docker"
   when:
   - openshift.common.is_containerized | bool
-  - not use_crio
+  - not l_use_crio
 
 - block:
   - name: Pull CLI Image
```

File: roles/openshift_version/tasks/set_version_containerized.yml [hunk 1]
```diff
@@ -42,12 +42,18 @@
   when:
   - openshift_version is defined
   - openshift_version.split('.') | length == 2
+  - not use_crio
```
Comment (ashcrow) [inline roles/openshift_version/tasks/set_version_containerized.yml]: You may need to define the variable in ```openshift-ansible/roles/openshift_version/tasks/main.yml```

File: roles/docker/tasks/systemcontainer_crio.yml [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+
+    - name: Set to default prepend
+      set_fact:
+        l_crio_image_prepend: "docker.io/gscrivano"
+        l_crio_image_name: "crio-o-fedora"
+
+    - name: Use Centos based image when distribution is Red Hat
+      set_fact:
+        l_crio_image_name: "cri-o-centos"
+      when: ansible_distribution == 'RedHat'
```
Comment (ashcrow) [inline roles/docker/tasks/systemcontainer_crio.yml]: Can we add ``CentOS`` here as well?

Comment (giuseppe) [inline roles/docker/tasks/systemcontainer_crio.yml]: sure, added

**Comment by giuseppe** [comment]
for node to be used as a system container, this change is required: https://github.com/openshift/origin/pull/15362

So that the CNI plugins are copied to the host and are made usable for CRI-O

**Comment by ashcrow** [comment]
I'm getting errors trying to install the system container on rhel non AH:

```
fatal: [192.168.122.3]: FAILED! => {
    "changed": false, 
    "failed": true, 
    "invocation": {
        "module_args": {
            "image": "docker.io/gscrivano/cri-o-centos:latest", 
            "name": "cri-o", 
            "state": "latest", 
            "values": []
        }
    }, 
    "rc": 1
}

MSG:

[Errno 2] No such file or directory: '/var/lib/containers/atomic/crio.0/rootfs/exports/hostfs/opt/cni/bin/bridge'
```

```
Failure summary:

  1. Host:     192.168.122.3
     Play:     Determine openshift_version to configure on first master
     Task:     docker : Install CRI-O System Container
     Message:  [Errno 2] No such file or directory: '/var/lib/containers/atomic/cri-o.0/rootfs/exports/hostfs/opt/cni/bin/bridge'
```

On the host itself:

```
# rpm -q atomic
atomic-1.17.2-9.git2760e30.el7.x86_64
# ls -la /var/lib/containers/atomic/cri-o.0/rootfs/exports/hostfs/opt/cni/bin/
total 0
drwxr-xr-x. 2 root root 161 Dec 31  1969 .
drwxr-xr-x. 3 root root  17 Dec 31  1969 ..
lrwxrwxrwx. 3 root root  23 Jul 28 14:40 bridge -> /usr/libexec/cni/bridge
lrwxrwxrwx. 3 root root  21 Jul 28 14:40 dhcp -> /usr/libexec/cni/dhcp
lrwxrwxrwx. 3 root root  24 Jul 28 14:40 flannel -> /usr/libexec/cni/flannel
lrwxrwxrwx. 3 root root  27 Jul 28 14:40 host-local -> /usr/libexec/cni/host-local
lrwxrwxrwx. 3 root root  23 Jul 28 14:40 ipvlan -> /usr/libexec/cni/ipvlan
lrwxrwxrwx. 3 root root  25 Jul 28 14:40 loopback -> /usr/libexec/cni/loopback
lrwxrwxrwx. 3 root root  24 Jul 28 14:40 macvlan -> /usr/libexec/cni/macvlan
lrwxrwxrwx. 3 root root  20 Jul 28 14:40 ptp -> /usr/libexec/cni/ptp
lrwxrwxrwx. 3 root root  23 Jul 28 14:40 sample -> /usr/libexec/cni/sample
lrwxrwxrwx. 3 root root  23 Jul 28 14:40 tuning -> /usr/libexec/cni/tuning
lrwxrwxrwx. 3 root root  21 Jul 28 14:40 vlan -> /usr/libexec/cni/vlan
```

The end result is that the container is exploded into ```/var/lib/containers/atomic/cri-o.0/``` but the symlink to ``cri-o`` is not created, and no service is added to systemd. Attempting to rerun the installer fails the same way.

Manual use of ``atomic`` shows the same error:

```
# atomic install --system-package=no --system --name=cri-o docker.io/gscrivano/cri-o-centos
Extracting to /var/lib/containers/atomic/cri-o.0
File /etc/crio/crio.conf already exists.
File /etc/crio/seccomp.json already exists.
[Errno 2] No such file or directory: '/var/lib/containers/atomic/cri-o.0/rootfs/exports/hostfs/opt/cni/bin/bridge'
```

Here is the traceback:

```
Traceback (most recent call last):
  File "/usr/bin/atomic", line 203, in <module>
    sys.exit(_func())
  File "/usr/lib/python2.7/site-packages/Atomic/install.py", line 132, in install
    return be.install(self.image, self.name)
  File "/usr/lib/python2.7/site-packages/Atomic/backends/_ostree.py", line 127, in install
    return self.syscontainers.install(image, name)
  File "/usr/lib/python2.7/site-packages/Atomic/syscontainers.py", line 223, in install
    return_value = self._install(image, name)
  File "/usr/lib/python2.7/site-packages/Atomic/syscontainers.py", line 363, in _install
    self._checkout(repo, name, image, 0, False, values=values, remote=self.args.remote, system_package=self.args.system_package)
  File "/usr/lib/python2.7/site-packages/Atomic/syscontainers.py", line 462, in _checkout
    system_package=system_package)
  File "/usr/lib/python2.7/site-packages/Atomic/syscontainers.py", line 763, in _do_checkout
    new_installed_files = RPMHostInstall.rm_add_files_to_host(installed_files, exports, prefix or "/", files_template=installed_files_template, values=values, rename_files=rename_files)
  File "/usr/lib/python2.7/site-packages/Atomic/rpm_host_install.py", line 64, in rm_add_files_to_host
    shutil.copy2(src_file, dest_path)
  File "/usr/lib64/python2.7/shutil.py", line 130, in copy2
    copyfile(src, dst)
  File "/usr/lib64/python2.7/shutil.py", line 82, in copyfile
    with open(src, 'rb') as fsrc:
IOError: [Errno 2] No such file or directory: '/var/lib/containers/atomic/cri-o.0/rootfs/exports/hostfs/opt/cni/bin/bridge'
```

**Comment by giuseppe** [comment]
@ashcrow thanks for trying that out!  I've pushed a fix for atomic here:

https://github.com/projectatomic/atomic/pull/1058

and a workaround for the CRI-O system container here:

https://github.com/projectatomic/atomic-system-containers/pull/95

**Comment by ashcrow** [comment]
@giuseppe excellent. I'll give it a quick go before calling it a day 😄

**Comment by giuseppe** [comment]
@ashcrow, let's make it easier :)  I will push the crio centos image to Docker Hub

**Comment by ashcrow** [comment]
@giuseppe perfect. You literally said that as I started down the rebuild path. 😆

**Comment by giuseppe** [comment]
@ashcrow pushed now!

**Comment by ashcrow** [comment]
It's installing cri-o and starting successfully now, but I'm getting errors with missing packages. I'm guessing repos are updating or something. I'll continue Monday when things settle down.

**Comment by mrunalp** [comment]
I am testing this on Centos. I had to install a copr repo for openvswitch. This is the latest issue I see:
```
Friday 28 July 2017  14:25:26 -0700 (0:00:00.305)       0:03:01.447 ***********
===============================================================================
docker : Install Docker ------------------------------------------------ 40.76s
docker : Pre-pull CRI-O System Container image ------------------------- 35.81s
os_firewall : need to pause here, otherwise the iptables service starting can sometimes cause ssh to fail -- 10.11s
docker : Ensure atomic is installed ------------------------------------- 6.42s
openshift_version : Get available atomic-openshift version -------------- 6.13s
openshift_excluder : Install docker excluder ---------------------------- 5.46s
openshift_excluder : Install openshift excluder ------------------------- 4.71s
os_firewall : Install iptables packages --------------------------------- 3.63s
docker : Start the Docker service --------------------------------------- 2.23s
docker : Install CRI-O System Container --------------------------------- 1.67s
openshift_docker_facts : Set docker facts ------------------------------- 1.43s
openshift_facts --------------------------------------------------------- 1.32s
openshift_docker_facts : Set docker facts ------------------------------- 1.31s
openshift_docker_facts : Set docker facts ------------------------------- 1.29s
openshift_health_check -------------------------------------------------- 1.29s
openshift_facts : Ensure various deps are installed --------------------- 1.23s
openshift_facts : Gather Cluster facts and set is_containerized if needed --- 0.96s
openshift_master_facts : Set master facts ------------------------------- 0.82s
docker : Set registry params -------------------------------------------- 0.82s
openshift_facts : Gather Cluster facts and set is_containerized if needed --- 0.77s

Failure summary:

  1. Host:     192.168.122.31
     Play:     Configure masters
     Task:     docker : Start the CRI-O service
     Message:  Unable to start service cri-o: Job for cri-o.service failed because the control process exited with error code. See "systemctl status cri-o.service" and "jou$nalctl -xe" for details.
```

**Comment by giuseppe** [comment]
@mrunalp is `/usr/share/rhel` used by CRI-O?  That bind mount is coming from the Docker container that we used to base the CRI-O on, if it is not used we can drop that. Otherwise I'll have to ensure it is present before runC starts

**Comment by ashcrow** [comment]
```/usr/share/rhel/secrets`` is provided by ```docker-latest```. We might as well create it if it doesn't exist.

```
$ sudo ls -dlZ /usr/share/rhel/secrets/
drwxr-x---. 2 root root system_u:object_r:usr_t:s0 4096 Jan 11  2017 /usr/share/rhel/secrets/
```

**Comment by mrunalp** [comment]
CRI-O shouldn't need it. It is probably from default mount list in our docker.

**Comment by giuseppe** [comment]
@ashcrow we might have a problem on Atomic Host since `/usr` is not writeable (so it is fine there until docker is installed on the host).  I am going to open a PR to drop the mount point, we can reconsider this later if CRI-O will need it.

**Comment by mrunalp** [comment]
Actually it depends on the rootfs and how you put it together. If it is RHEL based and uses yum to be built then we might need secrets from that directory.

**Comment by ashcrow** [comment]
@mrunalp thanks!

@giuseppe good thinking.

**Comment by giuseppe** [comment]
@mrunalp fix proposed here: https://github.com/projectatomic/atomic-system-containers/pull/97

**Comment by ashcrow** [comment]
Very close ....

Looks like ```system_images_registry``` is required, though I can't find documentation on what it should be set to (or a default).

@giuseppe do you know what it should be?
```
  1. Host:     192.168.122.3
     Play:     Configure masters
     Task:     openshift_cli : Pull CLI Image
     Message:  the field 'args' has an invalid value, which appears to include a variable that is undefined. The error was: 'dict object' has no attribute 'system_images_registry'
               
               The error appears to have been in '/home/steve/Tech/GITHUB/openshift-ansible/roles/openshift_cli/tasks/main.yml': line 26, column 5, but may
               be elsewhere in the file depending on the exact syntax problem.
               
               The offending line appears to be:
               
               - block:
                 - name: Pull CLI Image
                   ^ here
```

**Comment by giuseppe** [comment]
@ashcrow, that is the registry where to pull the system container images.  We should probably default it to `docker.io` for Fedora and CentOS and to `access.redhat.com` for RHEL.

**Comment by ashcrow** [comment]
@giuseppe got it. It's testing again, just very slowly ... had to restart it a few times.

**Comment by ashcrow** [comment]
cri-o ends up failing 

```
starting container process caused "process_linux.go:345: container init caused \"rootfs_linux.go:57: mounting \\\"/run/runc-ctrs/cri-o/notify.sock\\\" to rootfs \\\"/var/lib/containers/atomic/cri-o.0/rootfs\\\" at \\\"/var/lib/containers/atomic/cri-o.0/rootfs/run/systemd/notify\\\" caused \\\"no such file or directory\\\"\""
```

Verified that the path in ``rootfs`` does not exist.

**Comment by giuseppe** [comment]
@ashcrow are you testing on RHEL?  What version are you using?

**Comment by ashcrow** [comment]
Yes, RHEL 7.4 Installing with containerized=False.

**Comment by ashcrow** [comment]
The following commits from [my fork](https://github.com/ashcrow/openshift-ansible/commits/cri-o) allow for installation to continue based on our debugging:

- 695885b835444dcb1c5a441a19784d7c3d55b718
- ddd5afec7d3e48e36d1f5536a1b9446bdd4c79b9

Please cherry pick in.

**Comment by ashcrow** [comment]
My results get very close to working. We basically end up failing trying to verify the router deploys. However, ``cri-o`` is installed and running happily, the OpenShift master and node are running, etc.. I think this is about as far as we can go without starting to get into ``cri-o`` or router debugging.

@sdodson How much farther do you think we should continue in terms "verifying it works"?

**Comment by sdodson** [comment]
I've not yet found someone on @openshift/networking who could lend a hand, maybe this comment will do the trick.

File: roles/openshift_node/tasks/main.yml [hunk 3]
```diff
@@ -94,6 +94,13 @@
     - openshift.common.use_openshift_sdn | default(true) | bool
     - not openshift.common.is_containerized | bool
 
+- name: Restart cri-o
+  systemd:
+    name: cri-o
+    enabled: yes
+    state: restarted
+  when: openshift_docker_use_crio | default(false)
```
Comment (giuseppe) [inline roles/openshift_node/tasks/main.yml]: do you think we should move this at the end of the file and do only when `node_start_result | changed`?

Comment (ashcrow) [inline roles/openshift_node/tasks/main.yml]: We could. The reason I put it here though was that in my testing I always did it after the sdn was installed as that was what we needed more so than the node doing anything.

**Comment by sdodson** [comment]
https://github.com/openshift/origin/pull/14985 I'm told will help w/ CRI-O networking

**Comment by ashcrow** [comment]
Thanks @sdodson. Once that merges and is available I'll do a full test again.

**Comment by ashcrow** [comment]
I believe this was closed by the bot accidentally.

File: roles/docker/tasks/systemcontainer_crio.yml [hunk 4]
```diff
@@ -125,6 +125,12 @@
     src: crio.conf.j2
     backup: yes
 
+- name: Configure the CNI network
```
Comment (ashcrow) [inline roles/docker/tasks/systemcontainer_crio.yml]: I think we need the following above this task. On a fresh install the directory doesn't exist.

```yaml
- name: Ensure /etc/cni/net.d/ exists
  file:
    path: /etc/cni/net.d/
    state: directory
    mode: 0755
```

**Comment by giuseppe** [comment]
https://github.com/projectatomic/atomic-system-containers/pull/101 should fix the issues we saw with deploying containers

**Comment by ashcrow** [comment]
With https://github.com/projectatomic/atomic-system-containers/pull/101 this works!

**Comment by ashcrow** [comment]
@sdodson should we run the tests for these changes (to ensure we didn't break anything in the normal flow)?

**Comment by ashcrow** [comment]
bot, retest this please

**Comment by sdodson** [comment]
> @sdodson should we run the tests for these changes (to ensure we didn't break anything in the normal flow)?

yes please

**Comment by sdodson** [comment]
aos-ci-test

**Comment by sdodson** [comment]
[merge]

**Pull request merged**